text/0000-c-unwind-abi.md

```diff
@@ -0,0 +1,519 @@
+
+This change will be applied to all ABI strings other than `"Rust"`, such as
+`"system"`.
+
+## Interaction with `panic=abort`
+
+If a non-forced foreign unwind would enter a Rust frame via an `extern "C
+unwind"` ABI boundary, but the Rust code is compiled with `panic=abort`, the
+unwind will be caught and the process aborted.
+
+Conversely, non-forced unwinding from another language into Rust through an FFI
+entrypoint declared with `extern "C"` is always undefined behavior, and is not
+guaranteed to cause the program to abort under `panic=abort`. As noted
+[below][abi-boundaries-and-forced-unwinding], however, when compiling in debug
+mode, the compiler may be able to guarantee an abort in this case.
+
+`panic=abort` will have no impact on the behavior of forced unwinding.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## ABI boundaries and unforced unwinding
+[abi-boundaries-and-forced-unwinding]: #abi-boundaries-and-forced-unwinding
+
+This table shows the behavior of an unwinding operation reaching each type of
+ABI boundary (function declaration or definition). "UB" stands for undefined
+behavior. `"C"`-like ABIs are `"C"` itself but also related ABIs such as
+`"system"`.
+
+| panic runtime  | ABI          | `panic`-unwind                        | Unforced foreign unwind |
+| -------------- | ------------ | ------------------------------------- | ----------------------- |
+| `panic=unwind` | `"C unwind"` | unwind                                | unwind                  |
+| `panic=unwind` | `"C"`-like   | abort                                 | UB                      |
+| `panic=abort`  | `"C unwind"` | `panic!` aborts                       | abort                   |
+| `panic=abort`  | `"C"`-like   | `panic!` aborts (no unwinding occurs) | UB                      |
+
+In debug mode, the compiler could insert code to catch unwind attempts at
+`extern "C"` boundaries and `abort`; this would provide a safe way to discover
+(and fix) instances of this form of UB.
+
+## Frame deallocation and forced unwinding
+
+The interaction of Rust frames with C functions that deallocate frames (i.e.
+functions that may use forced unwinding on specific platforms) is independent
+of the panic runtime, ABI, or platform.
+
+* **When deallocating Rust non-POFs:** this is explicitly undefined behavior.
+* **When deallocating Rust [POFs][POF-definition]:** for now, this is not
+  specified, and must be considered undefined behavior. However, we do plan to
+  specify a safe way to deallocate POFs with `longjmp` or `pthread_exit` in [a
+  future RFC][unresolved-questions].
+
+## Additional limitations
+[additional-limitations]: #additional-limitations
+
+In order to limit the scope of this RFC, the following limitations are imposed:
+
+* No subtype relationship is defined between functions or function pointers
+  using different ABIs.
+* Coercions are not defined between `"C"` and `"C unwind"`.
```

Comment:
To me, "C unwind" looks like an "unwind" modifier being applied to the "C" ABI, or a combination of an ABI called "C" and an abi called "unwind". I think using a hypthen ("C-unwind") would make it clear that "C-unwind" and "C" are two distinct ABIs, which only differ in their handling of unwinding. It could also help prevent any confusion caused by `extern "unwind C"` and extern "unwind"` not being accepted.

Reply:
> an "unwind" modifier being applied to the "C" ABI

I mean, that's kind of what it is, no?

Reply:
It would avoid the need for the programmer to know that they're magic strings in this context.

We're used to seeing things like CSS class lists and command-line position argument lists which are space-separated lists of tokens, possibly unordered.

If there are no spaces in any of the valid ABI names, then there's no opportunity for anyone's intuition to get that point wrong.

Reply:
Sorry, the project group does not see any real disadvantage in the version with a space. If someone does write the wrong string, they'll get a compile error alerting them to the fact that `"unwind C"` is not a valid ABI string, and the fix will be trivial.

Personally, I do not see an advantage in "avoid[ing] the need for the programmer to know that they're magic strings." They are necessarily magic strings with or without the space, and I don't see why this knowledge makes FFI any harder.

Reply:
Your choice. It still feels like a needless "because we can" decision to use a space when, elsewhere in the language or the OS, snake case would be the typical choice.

The advantage I see is that the programmer is more likely to get it right on the first try, saving time and effort.

Reply:
I would also expect this to be hyphenated, similar to other magic strings like target triplets. I think stylistically I'd prefer for ABIs names not to contain spaces.

Reply:
(To be clear I'm personally not invested in this question one bit, but) for the record, it seems inevitable that someone, somewhere would eventually interpret a hyphen as subtraction.

Reply:
> (To be clear I'm personally not invested in this question one bit, but) for the record, it seems inevitable that someone, somewhere would eventually interpret a hyphen as subtraction.

what about `"C_unwind"` or `"C+unwind"`?

Reply:
`"C+unwind"` would give an even stronger impression that it's an unordered list of tokens than `"C unwind"` because, intuitively, it would seem like you're trying to draw a similarity to trait constraints.

Reply:
> `"C+unwind"` would give an even stronger impression that it's an unordered list of tokens than `"C unwind"` because, intuitively, it would seem that you're trying to draw a similarity to trait constraints.

yeah, makes sense

Reply:
That's also my issue with `"C unwind"`.

That pattern is used for things like passing an unordered list of space-separated tokens to some secondary parser, like some kind of `--compiler-args "-fdata-sections -ffunction-sections -Wl,--gc-sections"` argument to build automation or passing a `class="foo bar"` string to a templating engine.

Reply:
After further discussion, the project group does not have a strong opinion, and the majority view seems to be in favor of the hyphen. I have therefore changed all instances of `"C unwind"` to `"C-unwind"`.